### PR TITLE
libyang2: Make sysrepo-python works with libyang2

### PR DESCRIPTION
**Pull request by jeanseb6wind**
Since libyang2 migration, sysrepo-python is no more compatible with the
master branch of sysrepo. It's because sysrepo migrate to libyang2.

This patch adds the support of the new version of sysrepo.

### CODE REVIEW

**Comment by jeanseb6wind** [comment]
I have a problem with the last `sysrepo-python` commit on `master`.
Indeed, this commit uses the functions 

```
uint32_t sr_session_get_event_nc_id(sr_session_ctx_t *);
const char *sr_session_get_event_user(sr_session_ctx_t *);
```

Theses functions are not here anymore and now if you want to get theses data you have to use the new function 

```
int sr_session_get_orig_data(sr_session_ctx_t *session, uint32_t idx, uint32_t *size, const void **data)
```

You can compare this test to see the differences:

v1: https://github.com/sysrepo/sysrepo/blob/libyang1/tests/test_operational.c#L1082
v2: https://github.com/sysrepo/sysrepo/blob/master/tests/test_operational.c#L1001

The problem is that the `nc_id` and `user` should manually be set with `sr_session_push_orig_data`.

It's similar in v1, the `nc_id` should be set manually with `sr_session_set_nc_id(st->sess, 64);`. So I don't understand how works the last commit of `sysrepo-python` on `master`.

**Comment by jeanseb6wind** [comment]
Errors comes (in part) form the uses of bad version of libyang and sysrepo.

**Comment by tarkesh** [comment]
> Errors comes (in part) form the uses of bad version of libyang and sysrepo.

looks like we need the libyang-python 2.0 version too - https://github.com/pepa-cz/libyang-python/tree/libyang2_support which is still in a branch. Once that is installed, the check passes.
The syrepo cffi bindings are still on 1.x version somehow - i'm still looking for where they are defined. 
Did you have any luck in building it?

File: Makefile [hunk 1]
```diff
@@ -7,7 +7,7 @@ lint:
 	tox -e lint
 
 tests:
-	tox -e py3-libyang1
+	tox -e py37
```
Comment (rjarry) [inline Makefile:10]: This should not be pinned to a specific version of python.

File: sysrepo/session.py [hunk 2]
```diff
@@ -110,40 +121,107 @@ def set_error(self, xpath: Optional[str], message: str):
...
+        check_call(
+            lib.sr_session_push_orig_data, self.cdata, ffi.sizeof(c_user), p_user
         )
 
     def get_netconf_id(self) -> int:
         """
         It can only be called on an implicit sysrepo.Session (i.e., it can only be
-        called from an event callback)
+        called from an event callback).
+        It only works with netopeer2.
```
Comment (rjarry) [inline sysrepo/session.py:184]: I am not sure about this comment. Why would it only work with netopeer2? It works as soon as someone reports this via `sr_session_push_orig_data`.

File: sysrepo/session.py [hunk 3]
```diff
@@ -110,40 +121,107 @@ def set_error(self, xpath: Optional[str], message: str):
...
+        check_call(lib.sr_session_get_orig_data, self.cdata, 0, size, p_nc_id)
+
+        nc_id = ffi.cast("uint32_t *", p_nc_id[0])
+        return nc_id[0]
 
     def get_user(self) -> str:
         """
         It can only be called on an implicit sysrepo.Session (i.e., it can only be
         called from an event callback)
+        It only works with netopeer2.
```
Comment (rjarry) [inline sysrepo/session.py:208]: Same remark than above. Please remove this line.

File: sysrepo/session.py [hunk 4]
```diff
@@ -110,40 +121,107 @@ def set_error(self, xpath: Optional[str], message: str):
...
 
         :returns: the effective username of the event originator sysrepo session
         """
         if not self.is_implicit:
             raise SysrepoUnsupportedError("can only report user on implicit sessions")
-        return c2str(lib.sr_session_get_event_user(self.cdata))
+
+        if self.get_originator_name() != "netopeer2":
+            raise SysrepoUnsupportedError("can only report netconf id for netopeer2")
```
Comment (rjarry) [inline sysrepo/session.py:216]: I don't understand this. Why would you restrict this to a specific originator?

File: sysrepo/session.py [hunk 5]
```diff
@@ -833,19 +917,22 @@ def get_data_ly(
...
-            dnode_p,
+            sr_data_p,
         )
-        if not dnode_p[0]:
+        if not sr_data_p[0]:
             raise SysrepoNotFoundError(xpath)
-        return libyang.DNode.new(self.get_ly_ctx(), dnode_p[0]).root()
+        dnode = libyang.DNode.new(self.get_ly_ctx(), sr_data_p[0].tree).root()
+        # customize the free method to use the sysrepo free
+        dnode.free = lambda: lib.sr_release_data(sr_data_p[0])
```
Comment (rjarry) [inline sysrepo/session.py:934]: This is too much of a hack. Please create a subclass of `libyang.DNode` if you need to override something.

File: sysrepo/session.py [hunk 6]
```diff
@@ -1114,11 +1184,11 @@ def rpc_send_ly(
...
         in_dnode = ffi.cast("struct lyd_node *", rpc_input.cdata)
-        out_dnode_p = ffi.new("struct lyd_node **")
-        check_call(lib.sr_rpc_send_tree, self.cdata, in_dnode, timeout_ms, out_dnode_p)
-        if not out_dnode_p[0]:
+        sr_data_p = ffi.new("sr_data_t **")
+        check_call(lib.sr_rpc_send_tree, self.cdata, in_dnode, timeout_ms, sr_data_p)
+        if not sr_data_p[0]:
             raise SysrepoInternalError("sr_rpc_send_tree returned NULL")
-        return libyang.DNode.new(self.get_ly_ctx(), out_dnode_p[0])
+        return libyang.DNode.new(self.get_ly_ctx(), sr_data_p[0].tree)
```
Comment (rjarry) [inline sysrepo/session.py:1191]: Use the same class with overridden `free()`?

File: tests/test_subs_module_change.py [hunk 7]
```diff
@@ -198,15 +184,19 @@ def module_change_cb(event, req_id, changes, private_data):
                     },
                 }
             }
+            # sysrepo v1 behavior
+            # expected_changes = [
+            #    sysrepo.ChangeMoved(
+            #        "/sysrepo-example:conf/network/interface[name='eth2']",
+            #        after="[name='eth0']",
+            #    ),
+            # ]
```
Comment (rjarry) [inline tests/test_subs_module_change.py:193]: Remove this comment?

**Review by rjarry — CHANGES_REQUESTED** [review]
Hey there. There are a few changes that I would like. Thanks for working on this. :+1:

File: sysrepo/session.py (hunk 5 above)
Comment (jeanseb6wind) [reply sysrepo/session.py:929]: Hello @rjarry The problem is that the object is generate inside libyang. So I could update libyang by adding a parameter allowing to customize the free method but I'm not sure it's a good way to update libyang for that. What do you think about it ?

**Review by rjarry — APPROVED** [review]

**Comment by rjarry** [comment]
I merged this manually. Thanks a lot for your work!

**Pull request closed without merging**